sdk/core/core-rest-pipeline/src/policies/retryPolicy.ts

```diff
@@ -0,0 +1,130 @@
+          response = await next(request);
+          retryPolicyLogger.info(
+            `Retry ${retryCount}: Received a response from request`,
+            request.requestId
+          );
+        } catch (e) {
+          retryPolicyLogger.info(
+            `Retry ${retryCount}: Received an error from request`,
+            request.requestId
+          );
+          responseError = e as RestError;
+          if (responseError.name !== "RestError") {
+            throw responseError;
+          }
+          response = responseError.response;
+        }
+
+        retryPolicyLogger.info(
+          `Retry ${retryCount}: Processing ${strategies.length} retry strategies.`
+        );
+
+        for (const [i, strategy] of strategies.entries()) {
+          const strategyLogger = strategy.logger || retryPolicyLogger;
+          strategyLogger.info(`Retry ${retryCount}: Processing retry strategy ${strategy.name}.`);
+          let state: RetryStrategyState = {
+            ...strategyState[i],
+            retryCount,
+            response,
+            responseError
+          };
+
+          if (state.retryCount >= (state.maxRetries || DEFAULT_MAX_RETRIES)) {
+            strategyLogger.info(
+              `Maximum retries reached. Returning the last received response, or throwing the last received error.`
+            );
+            if (response) {
+              return response;
+            }
+            throw responseError;
+          }
+
+          if (!strategy.meetsConditions || strategy.meetsConditions(state)) {
+            state = strategy.updateRetryState(state);
+          } else {
+            strategyLogger.error(`Retry ${retryCount}: Does not meet conditions.`);
+            delete state.retryAfterInMs;
+          }
+          strategyState[i] = state;
+
+          if (request.abortSignal?.aborted) {
+            strategyLogger.error(`Retry ${retryCount}: Request aborted.`);
+            const abortError = new AbortError();
+            throw abortError;
+          }
+
+          if (state.throwError) {
+            strategyLogger.error(
+              `Retry ${retryCount}: Retry strategy ${strategy.name} throws error:`,
+              state.throwError
+            );
+            throw state.throwError;
+          }
+
+          if (state.retryAfterInMs) {
+            strategyLogger.info(
+              `Retry ${retryCount}: Retry strategy ${strategy.name} retries after ${state.retryAfterInMs}`
+            );
+            await delay(state.retryAfterInMs);
+            continue retryRequest;
+          }
+
+          if (state.redirectTo) {
+            strategyLogger.info(
+              `Retry ${retryCount}: Retry strategy ${strategy.name} redirects to ${state.redirectTo}`
+            );
+            request.url = state.redirectTo;
+            continue retryRequest;
+          }
+        }
+        retryPolicyLogger.info(`No retry strategy left. Returning the last received response.`);
+        if (response) {
```

Comment:
what do we do is `response` is falsy? It feels like we should guarantee we either exit this loop or throw. Otherwise, you could trivially infinite loop this by creating a retry policy without any strategies and a `next` function that always throws

Reply:
I think I fixed it here: https://github.com/Azure/azure-sdk-for-js/pull/19078/commits/a7dc30898262e60d3dac83ea95d3ea6cf46bef73